jwt.js

```diff
@@ -1,61 +1,114 @@
+    let {secret, resultProperty, ...rest} = options
+    steed.waterfall([
+      function getSecret (callback) {
+        secretCallback({}, {}, callback)
+      },
+      function sign (secret, callback) {
+        JWT.sign(request.body, secret, rest, function (err, token) {
+          if (err) {
+            callback(err)
+          } else {
+            callback(null, token)
+          }
+        })
+      }
+    ], function (err, result) {
+      if (_resultProperty) {
+        reply[_resultProperty] = result
+      } else {
+        request[_requestProperty] = result
+      }
+      return next(err, result)
+    })
+  } // end sign
 
-    if (typeof callback === 'function') {
-      JWT.sign(payload, secret, options, callback)
-    } else {
-      return JWT.sign(payload, secret, options)
-    }
-  }
+  function verify (request, reply, next) {
+    let token
+
+    if (request.headers && request.headers.authorization) {
+      const parts = request.headers.authorization.split(' ')
+      if (parts.length === 2) {
+        const scheme = parts[0]
+        const credentials = parts[1]
 
-  function verify (token, options, callback) {
-    assert(token, 'missing token')
-    assert(secret, 'missing secret')
-    options = options || {}
-    if (typeof options === 'function') {
-      callback = options
-      options = {}
+        if (/^Bearer$/i.test(scheme)) {
+          token = credentials
+        } else {
+          if (credentialsRequired) {
+            return next(new Error('Format is Authorization: Bearer [token]'))
+          } else {
+            return next()
+          }
+        }
+      } else {
+        return next(new Error('Format is Authorization: Bearer [token]'))
+      }
     }
 
-    if (typeof callback === 'function') {
-      JWT.verify(token, secret, options, callback)
-    } else {
-      return JWT.verify(token, secret, options)
+    if (!token) {
+      if (credentialsRequired) {
+        return next(new Error('No authorization token was found'))
+      } else {
+        return next()
+      }
     }
-  }
 
-  function decode (token, options) {
-    assert(token, 'missing token')
-    options = options || {}
-    return JWT.decode(token, options)
-  }
+    let decodedToken = JWT.decode(token, { complete: true }) || {}
+
+    steed.waterfall([
+      function getSecret (callback) {
+        secretCallback(request, decodedToken, callback)
+      },
+      function verify (secret, callback) {
+        JWT.verify(token, secret, options, function (err, decoded) {
+          if (err) {
+            callback(err)
+          } else {
+            callback(null, decoded)
+          }
+        })
+      }
+    ], function (err, result) {
+      if (_resultProperty) {
+        reply[_resultProperty] = result
+      } else {
+        request[_requestProperty] = result
+      }
+      return next(err, result)
+    })
+  } // end verify
 }
 
-module.exports = fp(fastifyJWT, '>=0.13.1')
+module.exports = fp(fastifyJwt, '0.x')
```

Comment:
can you make this `>= 0.39`.

Reply:
I think updating `fastify` from 0.13 to 0.39 broke the unit tests.

This ran and passed in 0.13, but now fails stating its unfinished.
```js
test('fastify-jwt should expose jwt methods', t => {
  t.plan(2)
  const fastify = Fastify()
  fastify
    .register(jwt, { secret: 'supersecret' })
    .after(() => {
      t.ok(fastify.jwt.sign)
      t.ok(fastify.jwt.verify)
    })
})
```
I think the `.after` was removed. Was it replaced with something else? 

I tried doing this but it says `can't read property sign of undefined`
```js
test('fastify-jwt should expose jwt methods', t => {
  t.plan(2)
  const fastify = Fastify()
  fastify
    .register(jwt, { secret: 'supersecret' })
  t.ok(fastify.jwt.sign)
  t.ok(fastify.jwt.verify)
})
```

Reply:
I'm quite puzzled on how to test this now. I'm not sure why but after I register `jwt` the fastify instance still returns undefined for `fastify.jwt`.

Reply:
this would be async. place the tests within `.after()`.

Reply:
<img width="1315" alt="screen shot 2018-01-17 at 11 38 31 am" src="https://user-images.githubusercontent.com/16144158/35054622-fc86ab9e-fb7a-11e7-83a6-85915fb2c496.png">
😕

Reply:
we have removed the callback in register. Do not pass `t.error`.

Reply:
<img width="1221" alt="screen shot 2018-01-17 at 12 48 25 pm" src="https://user-images.githubusercontent.com/16144158/35058050-bd0ec000-fb84-11e7-9445-f599fce2f89c.png">

Still an issue. I'll commit what I have as well as the broken test file in case you want to test it for yourself.